### PR TITLE
`fixNFTokenReserve`: Throw error when NFT buyer does not meet reserve requirement

### PR DESCRIPTION
**Pull request by shawnxie999**
## High Level Overview of Change
Add checks in the NFTokenAcceptOffer transactor to check if OwnerCount changed, if it did, then check for the reserve requirement.

### Context of Change
Fixes a problem where someone could accept an NFT sell offer without needing reserve.

This problem, however, does not occur when buy offer is involved or in brokered mode.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

File: src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp
Comment (scottschurr) [inline src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp:370]: There's always the (remote) possibility that this `read()` will return an empty `shared_ptr`.  So my personal preference is to check for `nullptr` before proceeding.  Maybe something like this:
```
    auto const sleBuyer = view().read(keylet::account(buyer));
    if (!sleBuyer)
        return tecINTERNAL;
```

File: src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp
Comment (scottschurr) [inline src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp:516]: I was a bit bothered that the code inserted into `acceptOffer` and `doApply` was very close to identical.  So I did a bit of nosing about.  It looks to me that the code that actually transfers the token in `acceptOffer` and `doApply` could be a common helper function.  It might look like this:
```
TER
NFTokenAcceptOffer::transferNFToken(
    AccountID const& buyer, AccountID const& seller, uint256 const& nftokenID)
{
    auto tokenAndPage = nft::findTokenAndPage(view(), seller, nftokenID);

    if (!tokenAndPage)
        return tecINTERNAL;

    if (auto const ret = nft::removeToken(
            view(), seller, nftokenID, std::move(tokenAndPage->page));
        !isTesSuccess(ret))
        return ret;

    auto const sleBuyer = view().read(keylet::account(buyer));
    if (!sleBuyer)
        return tecINTERNAL;

    std::uint32_t const buyerOwnerCountBefore =
        sleBuyer->getFieldU32(sfOwnerCount);

    auto const insertRet =
        nft::insertToken(view(), buyer, std::move(tokenAndPage->token));

    // if fixNFTokenReserve is enabled, check if the buyer has sufficient
    // reserve to own a new object, if their OwnerCount changed.
    //
    // There was an issue where the buyer accepts a sell offer, the ledger
    // didn't check if the buyer has enough reserve, meaning that buyer can get
    // NFTs free of reserve.
    if (view().rules().enabled(fixNFTokenReserve))
    {
        // To check if there is sufficient reserve, we cannot use mPriorBalance
        // because NFT is sold for a price. So we must use the balance after
        // the deduction of the potential offer price. A small caveat here is
        // that the balance has already deducted the transaction fee, meaning
        // that the reserve requirement is a few drops higher.
        auto const buyerBalance = sleBuyer->getFieldAmount(sfBalance);

        auto const buyerOwnerCountAfter = sleBuyer->getFieldU32(sfOwnerCount);
        if (buyerOwnerCountAfter > buyerOwnerCountBefore)
        {
            if (auto const reserve =
                    view().fees().accountReserve(buyerOwnerCountAfter);
                buyerBalance < reserve)
                return tecINSUFFICIENT_RESERVE;
        }
    }

    return insertRet;
}
```
Consider replacing `checkBuyerReserve()` with a function like this.  Then you'll be able to reduce the size of both `acceptOffer` and `doApply` and increase confidence that they are both doing the same thing more often.

If you are curious, you can see an example of what I'm describing as the top-most commit here: https://github.com/scottschurr/rippled/commits/shawnx-nft-reserve

**Review by scottschurr — COMMENTED** [review]
Don't let the comments I left discourage you.  This is a great pull request.  The fix looks good.  I think you did an especially good job with the tests.

I did leave some comments for you to consider, but they are mostly about different ways to structure the same code.  Feel free to push back on any of them.

Nice job!

File: src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp
Comment (shawnxie999) [reply src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp]: Thanks for the suggestion! This looks much better

File: src/test/app/NFToken_test.cpp
Comment (shawnxie999) [reply src/test/app/NFToken_test.cpp:6824]: Thanks for the suggestion. I think I would prefer removing the structured binding completely since it's indeed unnecessary to have the other field :)

**Comment by shawnxie999** [comment]
@scottschurr thanks for the great suggestions! this PR is now ready for re-review

**Review by scottschurr — APPROVED** [review]
Changes look good.  It would still be nice to fix the parameter name in the unit test lambda, but it's not important enough to hold up this pull request.

**Comment by shawnxie999** [comment]
@scottschurr sorry my bad, i thought I changed it, but it was lost when i was doing refactors. Just updated 👍

**Review by scottschurr — APPROVED** [review]
:+1:

File: src/test/app/NFToken_test.cpp
Comment (ckeshava) [reply src/test/app/NFToken_test.cpp:6825]: @scottschurr very insightful comment 👍 It's funny that you suggested the struct idea, because I believe that's how lambda's are implemented under the hood by the compiler :))

File: src/test/app/NFToken_test.cpp
Comment (ckeshava) [inline src/test/app/NFToken_test.cpp:6953]: shouldn't Bob have at least `incReserve + drops(10) /* current transaction fee */ + drops(10) /* fees lost due to the previous transaction*/` ?

I don't understand why this is different from testcase in line 6888 ?

File: src/test/app/NFToken_test.cpp
Comment (scottschurr) [reply src/test/app/NFToken_test.cpp:7160]: We verify the behavior of unit tests with and without new amendments because we...

1. Don't know when (or if) the amendment will be approved.  We don't want to damage existing behavior while adding an amendment.
2. We like for the current version of rippled to process ledgers that are at least a couple of years old.  This is very helpful for ledger replay.  So we want to verify that the unit tests show old behavior when recent amendments are not enabled.

Hopefully that explanation makes some sense.

Comment (ckeshava) [reply src/test/app/NFToken_test.cpp:7160]: thanks for the explanation! I forgot about the amendment voting process :) 👍

File: src/test/app/NFToken_test.cpp
Comment (ckeshava) [reply src/test/app/NFToken_test.cpp:6953]: On that note, I found it easier to understand a simpler test case:
```
// Bob has enough reserve for the Transaction Fees (10 drops) and the reserve for a new NFTokenPage
env(pay(env.master, bob, incReserve + drops(10)));
env.close();
env(token::acceptSellOffer(bob, sellOfferIndex));
env.close();
 
BEAST_EXPECT(ownerCount(env, bob) == 1);
```

Comment (shawnxie999) [reply src/test/app/NFToken_test.cpp:6953]: > I don't understand why this is different from testcase in line 6888 ?

The test case in line 68888 is testing the precise reserve requirements in the unit of drops. While in this test, it is exercising the behavior when Bob accepts many NFTs.

> shouldn't Bob have at least incReserve + drops(10) /* current transaction fee */ + drops(10) /* fees lost due to the previous transaction*/ ?

In this test case, Bob is funded with acctReserve + XRP(1). 1 XRP is more than enough for the transaction fees needed in this unit test. So we only need to worry about the incremental reserve that are taken up. Note that in unit test, the incremental reserve for an object is 50XRP, while the transaction fee is only 0.00001XRP

File: src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp
Comment (ckeshava) [reply src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp:342]: Okay, I understand that. I was confused by the wording "... meaning that the reserve requirement is a few drops higher". Apart from this, I had the same understanding 👍

File: src/test/app/NFToken_test.cpp
Comment (ckeshava) [reply src/test/app/NFToken_test.cpp:6953]: ah okay, I missed the fact that Bob has 1 XRP funds 👍

File: src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp
Comment (ckeshava) [reply src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp:342]: I feel the circumstance would be more clear if we eliminate that portion of the comment, but it's just a suggestion.

**Review by ckeshava — APPROVED** [review]
thanks for your work @shawnxie999 , this looks nice

File: src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp
Comment (shawnxie999) [reply src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp:342]: I think it would still be valuable to leave, since even tho 10 drops is not a lot. But still might be useful to the difference for those who might be reading the code.

**Comment by intelliot** [comment]
Intend to merge this after 2.0.1 is released

**Comment by tequdev** [comment]
@shawnxie999 Thank you for fixing the issue.

**Comment by intelliot** [comment]
Some unit tests were renamed. @shawnxie999 to resolve conflicts (no hurry).

**Comment by intelliot** [comment]
@shawnxie999 could you have a look at the "6 lines in your changes are missing coverage" reported above?

**Comment by shawnxie999** [comment]
@intelliot these are tecInternal errors, which are guarded by the preclaim checks already. I dont think we need additional tests needed for these lines 👍

**Comment by intelliot** [comment]
- @sgramkumar reviewing for QA

**Pull request merged**
